### ISSUE TITLE
Function whose argument is a callable type will accept any function that matches the `__call__` signature

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#2512]
**Describe the bug**
If a function takes an argument whose type implements `__call__`, then it will also incorrectly accept functions that implement the same callable signature. In fact, the behavior described in  #1888 is back again =/

**To Reproduce**
```python3
#pyright: strict

from typing import Protocol

def f() -> int:
    return 456

class MyProtocol(Protocol):
    def __call__(self) -> int:
        return 123

    def do_something(self):
        pass

def takes_my_protocol(x: MyProtocol):
    x.do_something()



class MyClass():
    def __call__(self) -> int:
        return 123

    def do_something_else(self):
        pass

def takes_my_class(x: MyClass):
    x.do_something_else()


takes_my_protocol(f) # no error
takes_my_class(f) #no error
```
Note that if you change the signature of `MyClass.__call__` or `MyProtocol.__call__`, pyright will report an error complaining about a mismatch in the signature, but not on the fact that `f` is not of type `MyClass` or that it doesn't match `MyProtocol`.

**VS Code extension or command-line**
pyright extension version 1.1.182

**Comment by erictraut** [comment]
The `takes_my_class` case is definitely an unintended bug. Yikes! I'm surprised this made it through our unit tests. Clearly there's a hole in the tests that I need to plug here.

The `takes_my_protocol` case is due to an intentional change that I made several months ago based on a feature request and subsequent discussions. I'm trying to find the thread, but I'm struggling to do so now. I'll try to reconstruct the argument from memory.

In Python, functions are objects. They must implement `__call__`, but they can also have other attributes and methods. The only way to describe the type of such objects within the Python type system is by using a protocol. For example, if you have a function that has an attribute `foo` of type `int`, you could describe it with the following protocol definition:

```python
class Foo(Protocol):
    foo: int
    def __call__(self, x: int) -> None:
        pass
```

The problem is that this doesn't interact well with other types within the type system that are designed to describe functions. That includes types described by function declarations and with the special form `typing.Callable`. It was therefore requested that we don't enforce type compatibility for attributes other than `__call__`.

Here's the [unit test](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/tests/samples/callbackProtocol5.py) that I wrote for this scenario. It demonstrates the use case that led to this change.

Perhaps the best approach here is to enforce type compatibility with the protocol and tell users who want to store additional attributes on their function objects that they should use an explicit `cast` call to cast a callable to the custom protocol type. Thoughts?

**Comment by erictraut** [comment]
Here's an example of how `cast` would be used. This looks quite reasonable to me.

```python
P = ParamSpec("P")
R = TypeVar("R", covariant=True)

class SomeFunc(Protocol[P, R]):
    other_attribute: int
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        ...

def other_func(f: Callable[P, R]) -> SomeFunc[P, R]:
    converted = cast(SomeFunc, f)
    converted.other_attribute = 1
    return converted

@other_func
def some_func(x: int) -> str:
    ...

some_func(1)
some_func.other_attribute = 5
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.183, which I just published. It will also be included in the next release of pylance.
